File: 268. Missing Number/Solution.cs
// Hash Table: linear space
public class Solution {
    public int MissingNumber(int[] nums) {
        
        HashSet<int> uniqueNums = new HashSet<int>(nums); 
        for (int i = 0; i < nums.Length; ++i) {
            if(!uniqueNums.Contains(i)) return i;
        }
        return nums.Length;
    }
}

// Xor: constant space
public class Solution {
    public int MissingNumber(int[] nums) {
        
        int missingNumber = nums.Length;
        for (int i = 0; i < nums.Length; ++i) {
            missingNumber ^= nums[i]^i;
        }
        return missingNumber;
    }
}

File: 268. Missing Number/solution.py
# set

class Solution:
    def missingNumber(self, nums: List[int]) -> int:

        unique_nums = set(nums)
        for i in range(len(nums)):
            if i not in unique_nums:
                 return i
        
        return len(nums)

# bit manipulation
class Solution:
    def missingNumber(self, nums: List[int]) -> int:

        missingNumber = len(nums)
        for i in range(len(nums)):
            missingNumber ^= nums[i]^i
        
        return missingNumber
